lightning/src/ln/channel.rs

```diff
@@ -1283,16 +1310,27 @@ impl<Signer: Sign> Channel<Signer> {
 				OutboundHTLCState::AwaitingRemovedRemoteRevoke(_) => (false, "AwaitingRemovedRemoteRevoke"),
 			};
 
+			let preimage_opt = match htlc.state {
+				OutboundHTLCState::RemoteRemoved(OutboundHTLCOutcome::Success(p)) => p,
+				OutboundHTLCState::AwaitingRemoteRevokeToRemove(OutboundHTLCOutcome::Success(p)) => p,
+				OutboundHTLCState::AwaitingRemovedRemoteRevoke(OutboundHTLCOutcome::Success(p)) => p,
```

Comment:
In this state, we've already sent a CS with the HTLCs removed, and are just waiting on a remote RAA to "forget" the HTLC. IIUC, because the HTLC will be removed before we sign the next remote CS, this won't lead to entries in `sign_counterparty_commitment` that we don't actually need (though we may need to re-sign the same CS again in case of reconnect), but it may mean we include the preimage in a `validate_holder_commitment` call even though, from the pov of the signer, its been fully and completely removed everywhere.

For a 1st pass, I think this is fine, but we should definitely document in the signer interface that we can include some preimages that were already fully removed and we prefer to over-provide preimages rather than under-provide them - its up to the signer to figure out what to do with them.

Reply:
should we just say:

```
	/// NOTE: a preimage may be provided here even if it has been already provided previously
	/// in [`sign_counterparty_commitment`].
```

in `validate_holder_commitment`?

Reply:
Likely, yea, I would word it even more aggressively and just disclaim any guarantees that the preimages are at all relevant and aren't just random strings.